### PR TITLE
templates and doc fixes

### PR DESCRIPTION
**Pull request by serboctor**
- Add docs to all templates.
- Add base classes to unimplemented templates to be able to run zero-robot.
- Make the names consistent and add a conventions doc.
- Fixes to several templates (node, bootstrap, hardware_check, container).

### CODE REVIEW

File: templates/node/node.py [hunk 1]
```diff
@@ -179,50 +146,52 @@ def uninstall(self):
         #     etcd_cluster_service.delete()
 
     def monitor(self):
-        self.logger.info("monitor")
+        self.logger.info('monitor')
```
Comment (zaibon) [inline templates/node/node.py]: you delete this method, it's not used

File: templates/node/README.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
+## template: github.com/zero-os/0-templates/node/0.0.1
+
+### Description:
+This template is responsible for managing a zero-os node.
+
+### Schema:
+
+- `id`: The mac address of the management network with the colons removed. **optional**
```
Comment (zaibon) [inline templates/node/README.md]: id is documented here, but not part of the schema.capnp

File: templates/node/node.py [hunk 3]
```diff
@@ -59,69 +60,33 @@ def update_data(self, d):
                 break
```
Comment (zaibon) [inline templates/node/node.py]: not shown in this diff, but the signature of the `update_data` method must be `def update_data(self, data):`

File: templates/vm/README.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
+## template: github.com/zero-os/0-templates/vm/0.0.1
+
+### Description:
+This template is responsible for managing a vm on a zero-os node.
+
+### Schema:
+
+- `node`: the node to deploy the vm on.
+- `id`: ??
```
Comment (zaibon) [inline templates/vm/README.md]: remove if not used

File: templates/vm/README.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+NicType enum: 
+- `default` 
+- `vlan`
+- `vxlan`
+- `bridge`
+
+
+### Actions:
+- `install`: creates a a vm and the hypervisor on the node.
+- `uninstall`: destroys the and deletes the service from 0-robot and the node.
```
Comment (zaibon) [inline templates/vm/README.md]: missing a word ?

File: templates/vm/vm.py [hunk 6]
```diff
@@ -40,9 +43,15 @@ def hypervisor(self):
         return self._hypervisor
 
     def install(self):
-        HV_TEMPLATE = 'github.com/jumpscale/0-robot/hypervisor/0.0.1'
+        node_name = self.data['node']
+        nodes = self.api.services.find(name=node_name, template_uid=NODE_TEMPLATE_UID)
```
Comment (zaibon) [inline templates/vm/vm.py]: use `services.exists`

File: templates/zeroos_bootstrap/README.md [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    - `redis_addr`: the redis address of the node to remove.
+
+
+```yaml
+services:
+    - github.com/zero-os/0-templates/zeroos_bootstrap/0.0.1__bootstrap:
+          zerotierClient: 'main'
+          zerotierNetID: '12ac4a1e7122ed7a'
+
+actions:
```
Comment (zaibon) [inline templates/zeroos_bootstrap/README.md]: This is actually not needed. the bootstrap is a recurring action. so it get scheduled automatically

File: templates/zeroos_bootstrap/zeroos_bootstrap.py [hunk 8]
```diff
@@ -60,17 +62,20 @@ def _wait_member_ip(self, member):
 
         resp = self.zt.client.network.getMember(member['nodeId'], netid)
         member = resp.json()
-        while len(member['config']['ipAssignments']) <= 0:
+
+        for _ in range(20):
```
Comment (zaibon) [inline templates/zeroos_bootstrap/zeroos_bootstrap.py]: here we could have a forever loop and decorate the method with the timeout decorator : https://jumpscale.github.io/0-robot/api/zerorobot/template/decorator.m.html
I think that will make the logic of the function more readable and the timeout duration is going to be more explicit with the decorator

Comment (zaibon) [reply templates/zeroos_bootstrap/zeroos_bootstrap.py]: Same trick can be used when trying to ping the node

File: templates/container/container.py [hunk 9]
```diff
@@ -2,44 +2,69 @@
...
-        container = self.node.containers.create(self.name, self.data['flist'], hostname=None, mounts=None, nics=None,
-                                                host_network=False, ports=None, storage=None, init_processes=None,
-                                                privileged=False, env=None)
-        self.state.set('actions', 'installed', 'ok')
-        return container
+        if self.node_sal is None:
+            raise RuntimeError("no connection to the zero-os node")
 
-    def start(self):
+        container_sal = self.node_sal.containers.create(self.name, self.data['flist'], hostname=None,
```
[the remark this answers is not in the document]
Comment (serboctor) [reply templates/container/container.py]: this was moved from the node template. From what I understand supporting all the features is not the priority now, right?

File: templates/container/container.py [hunk 10]
```diff
@@ -2,44 +2,69 @@
...
+            return
+
+        self.state.check('actions', 'install', 'ok')
         print('starting %s' % self.name)
-        self.container_client.start()
-        self.state.set('actions', 'started', 'ok')
+        self.container_sal.start()
+        self.state.set('actions', 'start', 'ok')
+
+    def stop(self, node_name=None):
```
[the remark this answers is not in the document]
Comment (serboctor) [reply templates/container/container.py]: in the node reboot, it stops and starts all the containers. If we are going to remove the parent attribute, then the filtering of the containers can't depend on it. So I execute the action on all containers and send the node name, if the container doesn't belong to this node, it ignores the action.

[the remark this answers is not in the document]
Comment (zaibon) [reply templates/container/container.py]: ok

File: templates/vm/vm.py [hunk 11]
```diff
@@ -40,9 +43,15 @@ def hypervisor(self):
...
+        if not nodes:
+            raise RuntimeError("service for node {} does'nt exist".format(node_name))
+
+        self.api.services.names[node_name].state.check('actions', 'install', 'ok')
+
         data = {
-            'node': self.node.name,
+            'node': self.node_sal.name,
```
Comment (zaibon) [inline templates/vm/vm.py]: should be `'node': self.data['node'],`

File: templates/container/container.py [hunk 12]
```diff
@@ -2,44 +2,68 @@
...
+        self.state.check('actions', 'install', 'ok')
+        self.logger.info('starting %s' % self.name)
+        self.container_sal.start()
+        self.state.set('actions', 'start', 'ok')
+
+    def stop(self, node_name=None):
+        if node_name and self.data['node'] != node_name:
+            return
+
+        self.state.check('actions', 'start', 'ok')
```
Comment (zaibon) [inline templates/container/container.py]: here we check for the start action to be ok. but the container is already started in the install method.
So we could have flow were start is never called.
I think we should set the install and start state to ok in the install action

**Pull request merged**